### PR TITLE
[NETBEANS-5318] Prevent unnecessary Gradle model reloads

### PR DESCRIPTION
**Pull request by sdedic**
Originally submitted in PR #2764, but separated for easier integration into its own PR.

## Prevent unnecessary reloads
When the project problem is being fixed, the project model is loaded again, but then the model is thrown away and loaded once again (unnecessarily). In 22424fb I've changed the behaviour to reuse the model - if another action did not alter the model in the meantime.

## Synchronization fixes
The `project` variable is accessed from several threads, at least from RELOAD_RP and any random thread that calls `NbGradleProjectImpl` getters. I've tried to synchronize that so the reference is replaced, but the `loadProject` is still executed outside mutexes and events are fired in RELOAD_RP.

### CODE REVIEW

File: extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java
Comment (JaroslavTulach) [inline extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java:216]: Why is `result` tri-state? Can it be named `mismatchOfOldAndNewProject`, right?

File: extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java
Comment (JaroslavTulach) [inline extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java:225]: this could be `} else if` and then `result` can be just plain `boolean` I think.

File: extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java
Comment (JaroslavTulach) [inline extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java:211]: Returned value seems to be unused in all invocations.

File: extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java
Comment (JaroslavTulach) [inline extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java:284]: dump!? clear? release?

File: extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java
Comment (JaroslavTulach) [inline extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java:191]: typo

**Review by JaroslavTulach — APPROVED** [review]
Hard to say whether the new behavior is correct. It however seems better than the previous one.

File: extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java
Comment (sdedic) [reply extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java:215]: possibly. I was never strong in naming things.

File: extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java
Comment (sdedic) [reply extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java:225]: Originally I wanted to propagate the action to the caller: accepted (null), changed by other party (true), original valid, but replacement rejected (false).

File: extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java
Comment (sdedic) [reply extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java:284]: `dumpProject` name was not changed in this PR.

File: extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java
Comment (sdedic) [reply extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java:216]: tri-state was explained below. If seems reasonable I would just leave it as it is; javadoc documents all three possible return values.

File: extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java
Comment (sdedic) [reply extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java]: Fixed in 2177f44a16

**Comment by sdedic** [comment]
.... 26 days without review from the component owner; I remember there's a work underway that may cause conflicts; clashes - any ETA where the competing PR could be merged so I can finish this one ?

**Review by lkishalmi — CHANGES_REQUESTED** [review]
Well, I do not see answers on my comments, no new changes. Till I do not see the complete logic fits, I would not approve this PR.

File: extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java
Comment (sdedic) [reply extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java:216]: OK, so degrading after reload is permitted. That's broken with this implementation - will revisit. Thanks.

Comment (sdedic) [reply extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java:216]: @lkishalmi I've checked the code again, recalling what I was thinking about that month ago: the code was written in such a way, as there can be multiple `loadProject` calls (potentialy) that would lead to `impl.project` replacement. The logic in the `recordOrReloadProject` basically ensures that if a project has been replaced **since the loadProject started** (the caller needs to save a reference), the `impl.project` will not be replaced for a worse version. But this is just for **competing** reloads.

Imagine a user clicking on Resolve problems, which will not make the project just `FULL` (whatever reason), but at the same time the background reload task will load the project and reaches `FULL_ONLINE`. If the final calls to `recordOrReloadProject` are reordered (different parallel threads) in an unfortunate sequence, the FULL_ONLINE read `impl.project` first, then it would degrade to FULL unnecessarily. The method's logic prevents that: if two replace processes are competing (caller has recorded `project` instance and passed it), the 'better of them' wins. Similar to AtomicReference.compareAndSet().

To be conservative, I've enabled this logic just from `GradleProjectProblemProvider` call. The other invocations pass `null` for `original`, so the comparison is disabled. So Reload action, reload based on desired quality change will always replace the project: I did not want to affect those occurrences, since I don't fully understand their semantics. I thought that it would not be good if "priming build" actually degraded already well-loaded project.

We can chat over the IM as you suggested or continue writing here. The logic to solve competing operations may be unnecessary.

**Comment by sdedic** [comment]
Changes in this PR became obsolete after #2747 - must be redone from scratch.

**Pull request closed without merging**
